fairworkflows/fairworkflow.py

```diff
@@ -395,11 +404,36 @@ def execute(self, num_threads=1):
         logger.setLevel(logging.INFO)
         logger.handlers = [log_handler]
 
-        result = run_parallel(
-            self.noodles_promise, n_threads=num_threads, registry=serial.base, db_file='temp_prov_fw.db',
-            always_cache=True, echo_log=False)
+        # A local function defining the serializers in the registry
+        # If we want to allow remote execution with noodles, we will
+        # need to define this function differently, but for multithreaded
+        # execution (most likely what we'll stick with) this is fine.
+        def registry():
+            return serial.pickle() + serial.base()
+
+        # TODO: Currently we are dumping a lot to an sqlite db, that will slow the
+        # execution down a bit. In future we will either write our own version of
+        # the runner, to grab that extra prov without any db writing, or we will
+        # read it from the database and convert it to RDF then.
+        with TemporaryDirectory() as td:
+            temp_db_fname = Path(td) / 'retro_prov.db'
+            result = run_parallel(self.noodles_promise, n_threads=num_threads,
+                                  registry=registry, db_file=temp_db_fname,
+                                  always_cache=True, echo_log=False)
+
+        # Generate the retrospective provenance as a (nano-) Publication object
```

Comment:
I think it would be cleaner if we factor out a method for this.

Reply:
Good idea. Done